internal/engine/wazevo/backend/isa/arm64/lower_instr.go

```diff
@@ -563,6 +563,144 @@ func (m *machine) LowerInstr(instr *ssa.Instruction) {
+		mov := m.allocateInstr()
+		switch lane {
+		case ssa.VecLaneI8x16:
+			arr = vecArrangementB
+			if signed {
+				mov.asMovFromVecSignedFIXME(rd, rn, arr, vecIndex(index))
+			} else {
+				mov.asMovFromVec(rd, rn, arr, vecIndex(index))
+			}
+		case ssa.VecLaneI16x8:
+			arr = vecArrangementH
+			if signed {
+				mov.asMovFromVecSignedFIXME(rd, rn, arr, vecIndex(index))
+			} else {
+				mov.asMovFromVec(rd, rn, arr, vecIndex(index))
+			}
+		case ssa.VecLaneI32x4:
+			arr = vecArrangementS
+			if signed {
+				mov.asMovFromVecSignedFIXME(rd, rn, arr, vecIndex(index))
+			} else {
+				mov.asMovFromVec(rd, rn, arr, vecIndex(index))
+			}
+		case ssa.VecLaneI64x2:
+			arr = vecArrangementD
+			if signed {
+				mov.asMovFromVecSignedFIXME(rd, rn, arr, vecIndex(index))
+			} else {
+				mov.asMovFromVec(rd, rn, arr, vecIndex(index))
+			}
+
+		case ssa.VecLaneF32x4:
+			arr = vecArrangementS
+			mov.asVecMovElement(rd, rn, arr, vecIndex(0), vecIndex(index))
+
+		case ssa.VecLaneF64x2:
+			arr = vecArrangementD
+			mov.asVecMovElement(rd, rn, arr, vecIndex(0), vecIndex(index))
+
+		default:
+			panic("not done")
+		}
+
+		m.insert(mov)
+
+	case ssa.OpcodeInsertlane:
+		x, y, index, _, lane := instr.InsertLaneOpData()
+		rn := m.getOperand_NR(m.compiler.ValueDefinition(x), extModeNone)
+		rm := m.getOperand_NR(m.compiler.ValueDefinition(y), extModeNone)
+		rd := operandNR(m.compiler.VRegOf(instr.Return()))
+
+		mov0 := m.allocateInstr()
+		mov0.asFpuMov128(rd.nr(), rn.nr())
+		m.insert(mov0)
+
+		mov := m.allocateInstr()
+		switch lane {
+		case ssa.VecLaneI8x16:
+			mov.asMovToVec(rd, rm, vecArrangementB, vecIndex(index))
+		case ssa.VecLaneI16x8:
+			mov.asMovToVec(rd, rm, vecArrangementH, vecIndex(index))
+		case ssa.VecLaneI32x4:
+			mov.asMovToVec(rd, rm, vecArrangementS, vecIndex(index))
+		case ssa.VecLaneI64x2:
+			mov.asMovToVec(rd, rm, vecArrangementD, vecIndex(index))
+		case ssa.VecLaneF32x4:
+			mov.asVecMovElement(rd, rm, vecArrangementS, vecIndex(index), vecIndex(0))
+		case ssa.VecLaneF64x2:
+			mov.asVecMovElement(rd, rm, vecArrangementD, vecIndex(index), vecIndex(0))
+		}
+		m.insert(mov)
+
+	case ssa.OpcodeSwizzle:
+		x, y, lane := instr.Arg2WithLane()
+		rn := m.getOperand_NR(m.compiler.ValueDefinition(x), extModeNone)
+		rm := m.getOperand_NR(m.compiler.ValueDefinition(y), extModeNone)
+		rd := operandNR(m.compiler.VRegOf(instr.Return()))
+
+		arr := ssaLaneToArrangement(lane)
+
+		tbl1 := m.allocateInstr()
+		tbl1.asVecTbl(1, rd, rn, rm, arr)
+		m.insert(tbl1)
+
+	case ssa.OpcodeShuffle:
+		x, y, lane1, lane2 := instr.Arg2WithLanes()
+		rn := m.getOperand_NR(m.compiler.ValueDefinition(x), extModeNone)
+		rm := m.getOperand_NR(m.compiler.ValueDefinition(y), extModeNone)
+		rd := operandNR(m.compiler.VRegOf(instr.Return()))
+
+		vReg, wReg := v29VReg, v30VReg
+
+		movv := m.allocateInstr()
+		movv.asFpuMov128(vReg, rn.nr())
+		m.insert(movv)
+
+		movw := m.allocateInstr()
+		movw.asFpuMov128(wReg, rm.nr())
+		m.insert(movw)
```

Comment:
I am sort-of surprised this works as it is. Is it all right? We are using v29, v30 because the instruction wants the 2 regs to be consecutive.

Reply:
why it is surprising?

Reply:
where `wReg` is used btw

Reply:
I guess it's just luckily it's working for sloppy spectest but the reality is that you have to tell the regalloc that Tbl2 to use not only vReg but also it's neighboring v30 as I commented on the useKind of Tbl2

Reply:
you see, I had a hunch something was working out of luck 😬

Reply:
and this is definitely something I want to have a unit test (and e2e tests)!